src/sync/cond.go

```diff
@@ -0,0 +1,80 @@
+package sync
+
+import "internal/task"
+
+type Cond struct {
+	L Locker
+
+	unlocking *earlySignal
+	blocked   task.Stack
+}
+
+// earlySignal is a type used to implement a stack for signalling waiters while they are unlocking.
+type earlySignal struct {
+	next *earlySignal
+
+	signaled bool
+}
+
+func (c *Cond) trySignal() bool {
+	// Pop a blocked task off of the stack, and schedule it if applicable.
+	t := c.blocked.Pop()
+	if t != nil {
+		scheduleTask(t)
+		return true
+	}
+
+	// If there any tasks which are currently unlocking, signal one.
+	if c.unlocking != nil {
+		c.unlocking.signaled = true
+		c.unlocking = c.unlocking.next
+		return true
+	}
+
+	// There was nothing to signal.
+	return false
+}
+
+func (c *Cond) Signal() {
+	c.trySignal()
+}
+
+func (c *Cond) Broadcast() {
+	// Signal everything.
+	for c.trySignal() {
+	}
+}
+
+func (c *Cond) Wait() {
+	// Add an earlySignal frame to the stack so we can be signalled while unlocking.
+	early := earlySignal{
+		next: c.unlocking,
+	}
+	c.unlocking = &early
+
+	// Temporarily unlock L.
+	c.L.Unlock()
+	defer c.L.Lock()
+
+	// If we were signaled while unlocking, immediately complete.
+	if early.signaled {
+		return
+	}
+
+	// Remove the earlySignal frame.
+	if c.unlocking == &early {
+		c.unlocking = early.next
```

Comment:
This appears to be a special case of the loop below. IMHO removing this special case makes the code easier to read and should not result in any performance issues as the loop below will not even enter the first iteration.

Reply:
I disagree, but I changed it.

Reply:
(this was meant as a suggestion, not as a requirement)

My original idea was to replace this:

```go
	if c.unlocking == &early {
		c.unlocking = early.next
	} else {
		prev := c.unlocking
		for prev.next != &early {
			prev = prev.next
		}
		prev.next = early.next
	}
```

with this:

```go
	prev := c.unlocking
	for prev.next != &early {
		prev = prev.next
	}
	prev.next = early.next
```

but now I see that's not quite the same as it leaves `c.unlocking` unchanged. So I think the original code is easier to read that the modified version, it doesn't simplify the code as I hoped.